server/src/main/java/org/elasticsearch/index/search/QueryParserHelper.java

```diff
@@ -167,22 +125,25 @@ public static FieldMapper getFieldMapper(MapperService mapperService, String fie
             if (fieldSuffix != null && context.fieldMapper(fieldName + fieldSuffix) != null) {
                 fieldName = fieldName + fieldSuffix;
             }
-            FieldMapper mapper = getFieldMapper(context.getMapperService(), fieldName);
-            if (mapper == null) {
+            MappedFieldType ft = context.fieldMapper(fieldName);
+            if (ft == null) {
                 // Unmapped fields are not ignored
                 fields.put(fieldOrPattern, weight);
                 continue;
             }
-            if (acceptMetadataField == false && mapper instanceof MetadataFieldMapper) {
+            if (acceptMetadataField == false && ft.name().startsWith("_")) {
                 // Ignore metadata fields
```

Comment:
Would this break BWC for some users? Underscored fieldnames were always by convention "ours" but are we now being stronger on this (either with docs or enforcing naming conventions?)